src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -56,7 +63,27 @@ public override T[] Rent(int minimumLength)
 
             // Gettings here means we have too big of a request OR all the buckets from 
             // index through _buckets.Length are taken so we need to allocate a buffer on-demand.
-            return new T[Utilities.GetMaxSizeForBucket(index)];
+            buffer = new T[Utilities.GetMaxSizeForBucket(index)];
+            if (log.IsEnabled())
+            {
+                int maxLength = Utilities.GetMaxSizeForBucket(_buckets.Length);
+                if (buffer.Length > maxLength)
+                    log.BufferAllocated(
+                        Utilities.GetBufferId(buffer),
+                        buffer.Length,
+                        Utilities.GetPoolId(this),
+                        -1 /*no bucket for an on-demand allocated buffer*/,
+                        ArrayPoolEventSource.BufferAllocationReason.OverMaximumSize);
+                else
+                    log.BufferAllocated(
+                        Utilities.GetBufferId(buffer),
+                        buffer.Length,
+                        Utilities.GetPoolId(this),
+                        -1 /*no bucket for an on-demand allocated buffer*/,
+                        ArrayPoolEventSource.BufferAllocationReason.PoolExhausted);
+            }
```

Comment:
Nit: You could avoid most of this duplicated code by changing the if/else to just be:
```C#
log.BufferAllocated(
    Utilities.GetBufferId(buffer),
    buffer.Length,
    Utilities.GetPoolId(this),
    -1, // no bucket for an on-demand allocated buffer,
    buffer.Length > maxLength? 
        ArrayPoolEventSource.BufferAllocationReason.OverMaximum :
        ArrayPoolEventSource.BufferAllocationReason.PoolExhausted);
```
Doing so makes the code shorter but also helps to highlight what the actual difference is.